259 piet testdoku spec final v (#274)
Schmiddl99/experimenteverwaltung-i2: Dokumentationen/se2_documentations/Testdokumentaion.adoc

= Testdokumentation: {project-name}
Piet Gutsche <piet.gutsche@htw-dresden.de>
{localdatetime}
include::../_includes/default-attributes.inc.adoc[]

== Einleitung
Das vorliegende Dokument erläutert die eingesetzten Teststrategien, deren Organisation sowie dabei aufgetretene Probleme

== Organisation

> Zur Aufteilung der Verantwortlichkeit für die Tests orientierten wir uns an der Testpyramide. Das Fundament dieser Teststrategie waren Unit-Tests, welche vom Entwickler geschrieben und durchgeführt wurden. Für diese nutzen wir das bereits von unseren Vorgängern benutzte Ruby Testframework RSpec.
> https://rspec.info/

Reviewer: ```suggestion
Zur Aufteilung der Verantwortlichkeit für die Tests orientierten wir uns an der Testpyramide. Das Fundament dieser Teststrategie waren Unit-Tests, welche vom Entwickler geschrieben und durchgeführt wurden. Für diese nutzen wir das bereits von unseren Vorgängern benutzte Ruby Testframework https://rspec.info/[RSpec].

```

> Wir legten dafür fest, das nur getesteter Code in den Main-Branch übernommen werden darf. Des weiteren wurde festgelegt, das vor jedem Merge ein Codereview durchgeführt wird. Dabei sollte der Entwickler die wesentlichen erstellten Funktionen erläutern. Als Kriterium zum Bestehen des Reviews wurde festgelegt, das der Code von Reviewer verstanden werden muss. Zusätzlich wurde vor jedem Merge eines Moduls sämtliche Tests der Regressiontestsuite ausgeführt. 

Reviewer: ```suggestion
Wir legten dafür fest, dass nur getesteter Code in den Main-Branch übernommen werden darf. Des weiteren wurde festgelegt, dass vor jedem Merge ein Codereview durchgeführt wird. Dabei sollte der Entwickler die wesentlichen erstellten Funktionen erläutern. Als Kriterium zum Bestehen des Reviews wurde festgelegt, dass der Code vom Reviewer verstanden werden muss. Zusätzlich wurden vor jedem Merge eines Moduls sämtliche Tests der Regressionstestsuite ausgeführt. 
```

> Für alle anderen Tests also UI-Tests, Funktionstest und Systemtests war der Tester verantwortlich. Wir entschieden uns dafür wann immer mit sinnvollen Aufwand möglich automatisierte Tests zu nutzen. Dafür nutzten wir das oben benannte Framework RSpec und Capybara zur Browserautomatisierung.
> https://teamcapybara.github.io/capybara/

Reviewer: ```suggestion
Für alle anderen Tests also UI-Tests, Funktionstest und Systemtests war der Tester verantwortlich. Wir entschieden uns dafür, wann immer mit sinnvollen Aufwand möglich, automatisierte Tests zu nutzen. Dafür nutzten wir das oben benannte Framework RSpec und https://teamcapybara.github.io/capybara/[Capybara] zur Browserautomatisierung.
```

> Somit wurde im Laufe unserer Arbeit die Regressiontestsuite, welche rudimentär schon von unseren Vorgängern aufgebaut wurde, stark erweitert. Diese erlaubt zukünftigen Entwicklern fehlerhafte Auswirkungen von Änderungen zügig zu erkennen.

Reviewer: ```suggestion
Somit wurde im Laufe unserer Arbeit die Regressionstestsuite, welche rudimentär schon von unseren Vorgängern aufgebaut wurde, stark erweitert. Diese erlaubt zukünftigen Entwicklern fehlerhafte Auswirkungen von Änderungen zügig zu erkennen.
```

Neben diesen automatisierten Tests führten wir eine Reihe manueller Tests durch, welche die automatisierten Test in Bereichen ergänzen in denen eine Automatisierung nur mit sehr hohem Aufwand möglich wäre.

Zusätzlich zu den von uns durchgeführten Tests erstellten wir ein Usertestprotokoll was von den zukünftigen Nutzern des Systems abgearbeitet werden sollte.

== Verhalten bei entdeckten Fehlern

> Wir legten fest, dass entdeckte Fehler dem Entwickler gemeldet werden. Gemeinsam mit diesem sollte eine Kategorisierung in optische Unschönheit(z.B. Button etwas zu lang oder nicht richtig angeordnet), leichter Fehler und schwerer Fehler vorgenommen werden. Schwere Fehler sollten dazu führen, dass die aktuellen Arbeiten pausiert und der Fehler sofort behoben werden. Leichte Fehler sollten nach Abschluss der aktuellen Arbeit an einer Funktion oder einem Feauture behoben werden. Bei optischen Unschönheiten sollte in Rahmen der weekly diskutiert werden, ob diese behoben werden oder ob sie vertretbar wären (z.B wenn Aufwand zum beseitigen unverhältnismäßig groß wäre).

Reviewer: ```suggestion
Wir legten fest, dass entdeckte Fehler dem Entwickler gemeldet werden. Gemeinsam mit diesem sollte eine Kategorisierung in optische Unschönheit (z.B. Button etwas zu lang oder nicht richtig angeordnet), leichter Fehler und schwerer Fehler vorgenommen werden. Schwere Fehler sollten dazu führen, dass die aktuellen Arbeiten pausiert und der Fehler sofort behoben werden. Leichte Fehler sollten nach Abschluss der aktuellen Arbeit an einer Funktion oder einem Feature behoben werden. Bei optischen Unschönheiten sollte im Rahmen der Weeklys diskutiert werden, ob diese behoben werden oder ob sie vertretbar wären (z.B. wenn Aufwand zum Beseitigen unverhältnismäßig groß wäre).
```

== Beschreibung der Tests
Die Beschreibungen der erstellten Testcase finden sich unter

> https://github.com/Schmiddl99/experimenteverwaltung-i2/blob/main/Dokumentationen/test/test_cases.adoc

Reviewer: ```suggestion
https://github.com/Schmiddl99/experimenteverwaltung-i2/blob/main/Dokumentationen/test/test_cases.adoc.

```

> Für alle automatisierten Tests finden sich in diesem Dokument die grobe Beschreibung des Tests und dessen Ziel. Die genauen Werte und die genaue Durchführung wurde in Form von Kommentaren den jeweiligen Specfiles angefügt. Die Zurordnung von Testcase zu automatisierten Test geschieht über die TC-Nummer (zb TC01) Dies verhindert zum einen die doppelte Beschreibung der Tests. Zum anderen erlaubt die genaue Beschreibung im Code zukünftigen Enwicklern eine zügiges Verständnis dieses Codes sowie eine einfache Erweiterbarkeit.

Reviewer: ```suggestion
Für alle automatisierten Tests finden sich in diesem Dokument die grobe Beschreibung des Tests und dessen Ziel. Die genauen Werte und die genaue Durchführung wurde in Form von Kommentaren den jeweiligen Specfiles angefügt. Die Zuordnung von Testcase zu automatisiertem Test geschieht über die TC-Nummer (z.B. TC01) Dies verhindert zum einen die doppelte Beschreibung der Tests. Zum anderen erlaubt die genaue Beschreibung im Code zukünftigen Entwicklern ein zügiges Verständnis dieses Codes sowie eine einfache Erweiterbarkeit.
```

Zu finden sind diese unter 

> https://github.com/Schmiddl99/experimenteverwaltung-i2/tree/main/src/spec/features

Reviewer: ```suggestion
https://github.com/Schmiddl99/experimenteverwaltung-i2/tree/main/src/spec/features.

```
